firebase-appdistribution/test-app/src/main/java/com/googletest/firebase/appdistribution/testapp/ScreenshotDetectionFeedbackTrigger.kt

```diff
@@ -0,0 +1,226 @@
+  override fun onActivityStarted(activity: Activity) {}
+  override fun onActivityStopped(activity: Activity) {}
+  override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
+
+  companion object {
+    const val TAG = "ScreenshotDetectionFeedbackTrigger"
+
+    private val permissionToRequest =
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) READ_MEDIA_IMAGES
+      else READ_EXTERNAL_STORAGE
+    private val shouldCheckIfPending = Build.VERSION.SDK_INT >= 29
+    private val contentProjection =
+      if (shouldCheckIfPending)
+        arrayOf(MediaStore.Images.Media.DATA, MediaStore.MediaColumns.IS_PENDING)
+      else arrayOf(MediaStore.Images.Media.DATA)
+
+    @SuppressLint("StaticFieldLeak") // Reference to Activity is set to null in onActivityPaused
+    private var instance: ScreenshotDetectionFeedbackTrigger? = null
+
+    /**
+     * Initialize the screenshot detection trigger for this application.
+     *
+     * This should be called during [Application.onCreate]. [enable] should then be called when you
+     * want to actually start detecting screenshots.
+     *
+     * @param application the [Application] object
+     * @param infoTextResourceId resource ID of info text to show to testers before giving feedback
+     */
+    open fun initialize(application: Application, infoTextResourceId: Int) {
+      if (instance == null) {
+        val handlerThread = HandlerThread("AppDistroFeedbackTrigger")
+        handlerThread.start()
+        instance = ScreenshotDetectionFeedbackTrigger(infoTextResourceId, Handler(handlerThread.looper))
+        application.registerActivityLifecycleCallbacks(instance)
+      }
+    }
+
+    /**
+     * Start listening for screenshots, and start feedback when a new screenshot is detected.
+     *
+     * @throws IllegalStateException if [initialize] has not been called yet
+     */
+    open fun enable() {
+      instance?.also {
+        it.isEnabled = true
+        it.listenForScreenshots()
+      } ?: throw IllegalStateException(
+        "You must call initialize() in your Application.onCreate() before enabling screenshot detection"
+      )
+    }
+
+    /**
+     * Stop listening for screenshots.
+     *
+     * @throws IllegalStateException if [initialize] has not been called yet
+     */
+    open fun disable() {
+      instance?.also {
+        it.isEnabled = false
+        it.stopListeningForScreenshots()
+      } ?: throw IllegalStateException(
+        "You must call initialize() in your Application.onCreate() before enabling screenshot detection"
+      )
+    }
+
+    private fun isExternalContent(uri: Uri) =
+      uri.toString().matches("${MediaStore.Images.Media.EXTERNAL_CONTENT_URI}/\\d+".toRegex())
+
+    private fun isPermissionGranted(activity: Activity) =
+      ContextCompat.checkSelfPermission(activity, permissionToRequest) == PERMISSION_GRANTED
+
+    private fun info(msg: String) = Log.i(TAG, msg)
```

Comment:
not sure if these pull their weight, but if you want to keep them, be my guest

Reply:
The log lines were formatting into these big ugly chunks like this:

```
Log.i(
  TAG,
  "Big long log line with a lot of text that will be logged out in one long line of logs"
)
```

But actually I'm already violating the linter by one-lining these shortened methods so maybe I should just one-line the Log.x calls...

Reply:
You could statically import `i()` and `e()` and shorten `TAG` to `T`. I'm not sure if I'm actually serious about this 🤣

Reply:
🤣 
```kotlin
i(T, m)
```

Code Rating:
✅ Gets the job done
✅ Good SLOC
✅ No verbose
❌ Readability